scripts/validate_string_concatenation.py

```diff
@@ -1,129 +1,377 @@
 #!/usr/bin/env python3
 """
-GH #30454
+Unwanted patterns test cases.
 
-Check where there is a string that needs to be concatenated.
+The reason this file exist despite the fact we already have
+`ci/code_checks.sh`,
+(see https://github.com/pandas-dev/pandas/blob/master/ci/code_checks.sh)
 
-This is necessary after black formating,
-where for example black transforms this:
-
->>> foo = (
-...         "bar "
-...         "baz"
-...     )
-
-into this:
-
->>> foo = ("bar " "baz")
-
-Black is not considering this as an
-issue (see issue https://github.com/psf/black/issues/1051),
-so we are checking it here.
+is that some of the test cases are more complex/imposible to validate via regex.
+So this file is somewhat an extensions to `ci/code_checks.sh`
 """
 
 import argparse
 import os
 import sys
 import token
 import tokenize
-from typing import Generator, List, Tuple
+from typing import IO, Callable, Generator, List, Tuple
+
+FILE_EXTENSIONS_TO_CHECK: Tuple[str, ...] = (".py", ".pyx", ".pyx.ini", ".pxd")
+
+
+def _get_literal_string_prefix_len(token_string: str) -> int:
```

Comment:
Hmm what's the point of this function? Does it only ever return 0 or 1?

Reply:
#### Short version:
The point of this function is to get the prefix length from string literals, so I can know where the actual string starts.

---

#### Long version:

The received named tuple from generating the tokens, contains the "line" attribute, which returns the whole line of code as if it was a string for example for 

```python
value = 42
```

The received "line" value is: ```'value = 42'``` (note the wrapping single quotes).

The problem is that I don't know if the string will be wrapped with single or double quotes, so I might get this (as the value of "line"):

```python
"f'My age is {age}'"
```

or this:

```python
'f"My age is {age}"'
```

The function may return:

* 0 - if there is no string prefix

* 1 - if the string prefix is just one character (e.g ```f```, ```r```)

* 2 - if the string prefix is two letters (e.g ```fr```)

---

The function is used to determine where to start cutting the received string, in order to check only the "raw" string.